source/server/connection_handler_impl.cc

```diff
@@ -477,6 +478,35 @@ void ConnectionHandlerImpl::ActiveTcpListener::newConnection(
 
   auto transport_socket = filter_chain->transportSocketFactory().createTransportSocket(nullptr);
   stream_info->setDownstreamSslConnection(transport_socket->ssl());
+
+  if constexpr (Network::Win32SupportsOriginalDestination()) {
+    ASSERT(socket->addressProvider().localAddressRestored() &&
+               config_->direction() != envoy::config::core::v3::UNSPECIFIED,
+           "Envoy on Windows needs the traffic direction to be set for original destination "
+           "listeners.");
+    if (config_->direction() == envoy::config::core::v3::OUTBOUND &&
+        socket->addressProvider().localAddressRestored()) {
+      ENVOY_LOG(debug, "[Windows] Querying for redirect record for outbound listener");
+      unsigned long redirectRecordsSize = 0;
+      auto redirect_records = std::make_shared<Network::EnvoyRedirectRecords>();
+      auto status = socket->genericIoctl(
+          SIO_QUERY_WFP_CONNECTION_REDIRECT_RECORDS, NULL, 0, (uint8_t*)redirect_records->buf_ptr_,
+          sizeof(redirect_records->buf_ptr_), redirect_records->buf_size_);
+      if (status.rc_ != 0) {
+        ENVOY_LOG(debug,
+                  "closing connection: cannot broker connection to original destination "
+                  "[Query redirect record failed] with error {}",
+                  status.errno_);
+        return;
+      }
+      stream_info->filterState()->setData(
+          Network::RedirectRecordsFilterState::key(),
+          std::make_unique<Network::RedirectRecordsFilterState>(redirect_records),
+          StreamInfo::FilterState::StateType::Mutable,
```

Comment:
ReadOnly? I don't think this is ever changed on a connection, is it?

Reply:
I will try it out with `ReadOnly` althought I am not sure if it will work because the `IOCTL` api does not guarantee that it will not modify the buffer.

Reply:
Changed to readonly in 583512a9750f17a04e7c1de4d0a428588fda08a1